### DISCUSSION TITLE
Is it normal that svg has smoother edges than using canvas?

### DISCUSSION THREAD

**Discussion by mirabledictu** [steveruizok/perfect-freehand#24] [General]
I have this sample output from SVG rendering

<img width="291" alt="Screen Shot 2021-07-26 at 8 16 48 PM" src="https://user-images.githubusercontent.com/86201195/126987280-368786b3-cd64-4a79-b3a7-df80c148dc8d.png">

It looks really good, smooth, but when I tried Canvas instead:

<img width="262" alt="Screen Shot 2021-07-26 at 8 10 24 PM" src="https://user-images.githubusercontent.com/86201195/126986538-eae9f1b2-949e-4cbd-b1a6-c2974abf8ae1.png">

Same implementation but instead passing the path to 2d constructor:

```ts
const pathData = getSvgPathFromStroke(getStroke(point, options))
const myPath = new Path2D(pathData)
ctx.fill(myPath)
```

It's using default options btw

**Comment by steveruizok** [comment]
Hey, my guess from that screenshot is that the canvas is at a lower dpi than your screen. Google around for “hi dpi html canvas” and you should be able to solve the issue.

**Reply by mirabledictu** [reply]
After some research I ended up with this:

```typescript
const canvas = this.getCanvasElement()
const rect = canvas.getBoundingClientRect()
const dpr =  window.devicePixelRatio || 1

canvas.width = rect.width * dpr
canvas.height = rect.height * dpr
const ctx = canvas.getContext('2d')
ctx?.scale(dpr, dpr)

canvas.style.width = rect.width + 'px'
canvas.style.height = rect.height + 'px'
```

Tho the edges are still a bit rough, but changing the `dpr` to `2` made it look like this

![Screen Shot 2021-07-27 at 3 07 02 PM](https://user-images.githubusercontent.com/86201195/127110934-68fa877d-282c-4f9c-9be5-4479a6ef33a9.png)

Weird that `window.devicePixelRatio` gives `1` but `2` looks better...

**Reply by mirabledictu** [reply]
Also I make it perfect looking like svg by doing `ctx?.clearRect(0,0,canvas.width,canvas.height)`

**Reply by steveruizok** [reply]
Looks good! If you'd like to adapt the new example project to canvas rendering, I'd be happy to add it to the README. Otherwise I'll add it to my own todo list.
